File: sdk/python/examples/pagination_and_errors.py
"""
Pagination and Error Handling Examples

Demonstrates cursor-based pagination, offset pagination,
and error handling patterns with the Kokonut Python SDK.

Usage:
    export KOKONUT_API_URL=http://localhost:8055
    export KOKONUT_EMAIL=field@kokonut.local
    export KOKONUT_PASSWORD=field1234
    python sdk/python/examples/pagination_and_errors.py
"""

import os
import sys

sys.path.insert(0, os.path.join(os.path.dirname(__file__), ".."))

from kokonut import KokonutClient, ListOptions
from kokonut.client import (
    KokonutError,
    AuthenticationError,
    NotFoundError,
    PermissionError,
    ValidationError,
)


def main():
    client = KokonutClient(os.environ.get("KOKONUT_API_URL", "http://localhost:8055"))

    # --- Authentication with error handling ---
    email = os.environ.get("KOKONUT_EMAIL", "field@kokonut.local")
    password = os.environ.get("KOKONUT_PASSWORD", "field1234")

    try:
        client.login(email, password)
        print(f"Logged in as {email}")
    except AuthenticationError as e:
        print(f"Authentication failed: {e}")
        print("Check your email and password.")
        return
    except KokonutError as e:
        print(f"Login error: {e}")
        return

    # --- Page-based pagination ---
    print("\n--- Page-based Pagination (Locations) ---")
    page = 1
    page_size = 5
    all_locations = []

    while True:
        try:
            locations = client.locations.list(
                ListOptions(
                    page=page,
                    limit=page_size,
                    sort=["name"],
                    meta="total_count",
                )
            )

            if not locations:
                break

            all_locations.extend(locations)
            print(f"  Page {page}: {len(locations)} locations")

            if len(locations) < page_size:
                break
            page += 1

        except PermissionError as e:
            print(f"Permission denied on page {page}: {e}")
            break
        except KokonutError as e:
            print(f"Error fetching page {page}: {e}")
            break

    print(f"  Total: {len(all_locations)} locations")

    # --- Offset-based pagination (harvest events) ---
    print("\n--- Offset-based Pagination (Harvest Events) ---")
    offset = 0
    limit = 10
    total_fetched = 0

    while True:
        try:
            events = client.harvest_events.list(
                ListOptions(
                    offset=offset,
                    limit=limit,
                    sort=["-harvest_date"],
                )
            )

            if not events:
                break

            for event in events:
                total_fetched += 1
                print(f"  {event['harvest_date']}: {event['quantity']} {event['unit']}")

            if len(events) < limit:
                break
            offset += limit

        except ValidationError as e:
            print(f"Validation error: {e}")
            break
        except KokonutError as e:
            print(f"Error: {e}")
            break

    print(f"  Total fetched: {total_fetched}")

    # --- Filter with error handling ---
    print("\n--- Filtered Query (Sales with payment_status=pending) ---")
    try:
        pending_sales = client.sales_events.list(
            ListOptions(
                filter={"payment_status": {"_eq": "pending"}},
                sort=["-sale_date"],
                limit=5,
            )
        )
        for sale in pending_sales:
            print(f"  {sale['sale_date']}: ${sale['total_amount']} ({sale['payment_status']})")
    except NotFoundError:
        print("  No pending sales found.")
    except PermissionError as e:
        print(f"  Cannot access sales: {e}")
    except KokonutError as e:
        print(f"  Error: {e}")

    # --- Create with validation error handling ---
    print("\n--- Create with Error Handling ---")
    try:
        new_location = client.locations.create({
            "name": "Test Pagination Farm",
            "slug": "test-pagination-farm",
            "country": "Costa Rica",
            "status": "active",
        })
        print(f"  Created: {new_location['name']} ({new_location['id']})")

        # Try creating a duplicate (should fail gracefully)
        try:
            client.locations.create({
                "name": "Test Pagination Farm",
                "slug": "test-pagination-farm",  # duplicate slug
                "country": "Costa Rica",
                "status": "active",
            })
        except ValidationError as e:
            print(f"  Duplicate slug caught: {e}")
        except KokonutError as e:
            print(f"  Duplicate slug error: {e}")

    except ValidationError as e:
        print(f"  Validation error: {e}")
    except PermissionError as e:
        print(f"  Permission denied: {e}")
    except KokonutError as e:
        print(f"  Error: {e}")

    # --- Generic error handler ---
    print("\n--- Generic Error Handling Pattern ---")
    try:
        locations = client.locations.list(ListOptions(limit=3))
        print(f"  Fetched {len(locations)} locations")
    except AuthenticationError:
        print("  Session expired. Re-login required.")
    except PermissionError:
        print("  Insufficient permissions for this operation.")
    except NotFoundError:
        print("  Resource not found.")
    except ValidationError as e:
        print(f"  Invalid request: {e}")
    except KokonutError as e:
        print(f"  API error: {e}")
    except Exception as e:
        print(f"  Unexpected error: {e}")

    print("\nDone.")


if __name__ == "__main__":
    main()

File: sdk/javascript/examples/pagination-and-errors.ts
/**
 * Pagination and Error Handling Examples
 *
 * Demonstrates page-based pagination, offset pagination,
 * and error handling patterns with the Kokonut JS/TS SDK.
 *
 * Usage:
 *   KOKONUT_API_URL=http://localhost:8055 \
 *   KOKONUT_EMAIL=field@kokonut.local \
 *   KOKONUT_PASSWORD=field1234 \
 *   npx tsx sdk/javascript/examples/pagination-and-errors.ts
 */

import { KokonutClient, ListOptions } from '../src/index.js';

async function main() {
  const client = new KokonutClient(
    process.env.KOKONUT_API_URL || 'http://localhost:8055'
  );

  // --- Authentication with error handling ---
  const email = process.env.KOKONUT_EMAIL || 'field@kokonut.local';
  const password = process.env.KOKONUT_PASSWORD || 'field1234';

  try {
    await client.login(email, password);
    console.log(`Logged in as ${email}`);
  } catch (e: any) {
    console.error(`Authentication failed: ${e.message}`);
    return;
  }

  // --- Page-based pagination ---
  console.log('\n--- Page-based Pagination (Locations) ---');
  let page = 1;
  const pageSize = 5;
  const allLocations: any[] = [];

  while (true) {
    try {
      const locations = await client.locations.list({
        page,
        limit: pageSize,
        sort: ['name'],
        meta: 'total_count',
      } as ListOptions);

      if (!locations || locations.length === 0) break;

      allLocations.push(...locations);
      console.log(`  Page ${page}: ${locations.length} locations`);

      if (locations.length < pageSize) break;
      page++;
    } catch (e: any) {
      console.error(`  Error on page ${page}: ${e.message}`);
      break;
    }
  }

  console.log(`  Total: ${allLocations.length} locations`);

  // --- Offset-based pagination (harvest events) ---
  console.log('\n--- Offset-based Pagination (Harvest Events) ---');
  let offset = 0;
  const limit = 10;
  let totalFetched = 0;

  while (true) {
    try {
      const events = await client.harvestEvents.list({
        offset,
        limit,
        sort: ['-harvest_date'],
      } as ListOptions);

      if (!events || events.length === 0) break;

      for (const event of events) {
        totalFetched++;
        console.log(`  ${event.harvest_date}: ${event.quantity} ${event.unit}`);
      }

      if (events.length < limit) break;
      offset += limit;
    } catch (e: any) {
      console.error(`  Error at offset ${offset}: ${e.message}`);
      break;
    }
  }

  console.log(`  Total fetched: ${totalFetched}`);

  // --- Filter with error handling ---
  console.log('\n--- Filtered Query (Pending Sales) ---');
  try {
    const pendingSales = await client.salesEvents.list({
      filter: { payment_status: { _eq: 'pending' } },
      sort: ['-sale_date'],
      limit: 5,
    } as ListOptions);

    for (const sale of pendingSales) {
      console.log(`  ${sale.sale_date}: $${sale.total_amount} (${sale.payment_status})`);
    }
  } catch (e: any) {
    console.error(`  Error: ${e.message}`);
  }

  // --- Create with validation error handling ---
  console.log('\n--- Create with Error Handling ---');
  try {
    const newLocation = await client.locations.create({
      name: 'Test Pagination Farm',
      slug: 'test-pagination-farm-ts',
      country: 'Costa Rica',
      status: 'active',
    });
    console.log(`  Created: ${newLocation.name} (${newLocation.id})`);

    // Try duplicate (should fail gracefully)
    try {
      await client.locations.create({
        name: 'Test Pagination Farm',
        slug: 'test-pagination-farm-ts',
        country: 'Costa Rica',
        status: 'active',
      });
    } catch (e: any) {
      console.log(`  Duplicate slug caught: ${e.message}`);
    }
  } catch (e: any) {
    console.error(`  Create error: ${e.message}`);
  }

  // --- Generic error handler ---
  console.log('\n--- Generic Error Handling Pattern ---');
  try {
    const locations = await client.locations.list({ limit: 3 } as ListOptions);
    console.log(`  Fetched ${locations.length} locations`);
  } catch (e: any) {
    if (e.message?.includes('401')) {
      console.error('  Session expired. Re-login required.');
    } else if (e.message?.includes('403')) {
      console.error('  Insufficient permissions.');
    } else if (e.message?.includes('404')) {
      console.error('  Resource not found.');
    } else {
      console.error(`  API error: ${e.message}`);
    }
  }

  console.log('\nDone.');
}

main();
